### DISCUSSION TITLE
How could I use quill for a system that uses a plugin system?

### DISCUSSION THREAD

**Discussion by qiuwei** [odygrd/quill#532] [Q&A]
I am working on a mid-sized framework that employs a plugin system.
There is a clear c interface between the framework and plugins.

Currently, I create a logging interface like 
```
extern "C" void log_message(int level, const char* message);
```
The plugin calls the method to pass the message to the framework, which uses quill to do the actual logging.

However, this approach seems suboptimal because the plugin is required to do the formatting and there is char array copying involved in quill.

Any suggestions on this?

**Comment by odygrd** [comment]
Hey, yes with the above you are passing already formatted messages. There is another problem when not using macros for logging which is that the language will always evaluate the arguments for the function regardless of the log level used.

Would it be possible to setup/start the backend as part of the framework and then allow the plugins to call the log macros directly ?  Or are you wrapping the whole framework in a C wrapper and require all the plugins to only use C ?

**Reply by qiuwei** [reply]
The boundary between the framework and the plugins is only C but the plugins can use CPP.

I also thought about letting the framework set up the backend, but then the logging part as an implementation detail is coupled with the plugins.

The framework is actually a trading system. The plugins as strategies are supposed to work under the trading framework as well as in a backtesting engine. The backtesting engine might not use Quill for the logging.

**Reply by odygrd** [reply]
I can’t think of a way doing this with a c api. You can pass variadic arguments in c but then you have to use `va_list` which doesn’t keep the type information, so basically your plugins will end up calling something that looks like `printf`

One solution to support different log libraries would be your framework to define some macros in a header like

```
#define LOG_INFO(format, …) \
   QUILL_LOG_INFO(logger, format, ##__VA_ARGS__);
```

On the backrest codebase create a header with the same filename and define exactly the same macros 

```
#define LOG_INFO(format, …) \
  Pass format and args to your logging library 
```
